example/Binary.hs

```diff
@@ -0,0 +1,389 @@
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TypeFamilies #-}
+
+module Binary (
+  Bit (..),
+  LogicBit (..),
+  _LogicO,
+  _LogicI,
+  _LogicO',
+  _LogicI',
+  Binary,
+  zeroo,
+  poso,
+  gtlo,
+  binaryo,
+  addo,
+  subo,
+  lesso,
+  mulo,
+  divo,
+  example,
+) where
+
+import Control.Lens (Prism, from)
+import Control.Lens.TH (makePrisms)
+import Control.Monad (void)
+import Data.Bifunctor (bimap)
+import Data.Function ((&))
+import Data.Tagged (Tagged)
+import GHC.Generics (Generic)
+
+import Core
+import Goal
+import LogicalBase
+import Match
+import TH
+
+data Bit = O | I deriving (Eq, Show, Generic)
+
+makeLogic ''Bit
+deriving instance Show LogicBit
+
+makePrisms ''LogicBit
+
+_LogicO'
+  :: Prism
+      (Tagged (o, i) LogicBit)
+      (Tagged (o', i) LogicBit)
+      (Tagged o ())
+      (Tagged o' ())
+_LogicO' = from _Tagged . _LogicO . _Tagged
+
+_LogicI'
+  :: Prism
+      (Tagged (o, i) LogicBit)
+      (Tagged (o, i') LogicBit)
+      (Tagged i ())
+      (Tagged i' ())
+_LogicI' = from _Tagged . _LogicI . _Tagged
+
+type Binary = [Bit]
+
+instance Num Binary where
+  fromInteger 0 = []
+  fromInteger 1 = [I]
+  fromInteger x | x > 1 = bit : fromInteger (x `div` 2)
+   where
+    bit
+      | even x = O
+      | otherwise = I
+  fromInteger _ = error "Binary numbers cannot be negative"
+
+  (+) = undefined
+  (*) = undefined
+  abs = id
+  signum = undefined
+  negate = undefined
+
+zeroo :: Term Binary -> Goal ()
+zeroo n = n === Value LogicNil
+
+cons :: (Logical a) => Term [a] -> Goal (Term a, Term [a])
+cons list = do
+  (x, xs) <- fresh
+  list === Value (LogicCons x xs)
+  return (x, xs)
+
+poso :: Term Binary -> Goal ()
+poso = void . cons
+
+gtlo :: Term Binary -> Goal ()
+gtlo n = do
+  (_, ns) <- cons n
+  poso ns
+
+{- FOURMOLU_DISABLE -}
+binaryo :: Term Binary -> Goal ()
```

Comment:
This seems to check that (reversed) binary number does not have (leading) non-significant zeros. It also checks that bits are `I` or `O`, but we already know that from the type of `Bit`. Of course, to generate bits, we need to mention both. Still, it seems slightly inefficient to have both `poso` and `binaryo` since we sill traverse the same `LogicCons` twice (I do not have any suggestion here yet).

Also, in this case, we can actually use a custom type to enforce this, for example:

```haskell
data Binary
  = Zero       -- zero, corresponds to [] :: [Bit]
  | One [Bit]  -- non-zero, corresponds to (bs ++ [1]) :: [Bits]
```

In this case, we no longer have to _check_ correctness of binary numbers. Moreover, the relation generating all possible `Binary` values can be done using `Generic` in a standard way.

I wonder, if we can leverage any of that for a more efficient implementation.

Reply:
This definition indeed eliminates invalid values, but it makes working with the last bit more complicated, thus translating the original implementation becomes harder. While it may be worth exploring this definition later, I'm not sure it will bring any advantages. `genb` (which is `binaryo` in our speak) is not used anywhere in the original implementation. I added one myself here:

https://github.com/SnejUgal/static-minikanren/blob/e693a4fca7d9b537646c894d87529e8d19bf7124/example/Binary.hs#L263-L273

but I did it only because one of the solutions to `divo n m p q` was `n == p == [_], q == [I]`, which clearly begs to try putting `n == [0]`. not sure if adding this line was worth it though.